runtime/vm/resolvesupport.cpp

```diff
@@ -61,7 +61,9 @@ finalFieldSetAllowed(J9VMThread *currentThread, bool isStatic, J9Method *method,
 		if (romClass->majorVersion >= 53) {
 			J9ROMMethod *romMethod = J9_ROM_METHOD_FROM_RAM_METHOD(method);
 			J9UTF8 *name = J9ROMMETHOD_NAME(romMethod);
-			if (('<' != J9UTF8_DATA(name)[0]) || (J9UTF8_LENGTH(name) != strlen(isStatic ? "<clinit>" : "<init>"))) {
+			if (('<' != J9UTF8_DATA(name)[0])
+					|| (J9UTF8_LENGTH(name) != strlen(isStatic ? "<clinit>" : "<init>"))
+					|| ((fieldClass != methodClass) && (J9ROMCLASS_IS_UNSAFE(methodClass->romClass)))) {
```

Comment:
Won't the logic here treat `(fieldClass != methodClass) == true` and `J9ROMCLASS_IS_UNSAFE(methodClass->romClass) == true` as an error case?

Shouldn't that be `J9ROMCLASS_IS_UNSAFE(methodClass->romClass) == 0`?

Reply:
From what I understand, the existing visibility checks called for final field access prior to this use are:
```
checkResult = checkVisibility(vmStruct, classFromCP, definingClass, J9_JAVA_PRIVATE, resolveFlags);
if (checkResult < J9_VISIBILITY_ALLOWED) { ... }
```
And the access checking within vm/visible.c uses the order:
```
UDATA result = J9_VISIBILITY_ALLOWED;
...
if (J9ROMCLASS_IS_UNSAFE(sourceClass->romClass) == 0) {
...
if (modifiers & J9AccPrivate) {
	/* Private */
	if ((sourceClass != destClass)
#if defined(J9VM_OPT_VALHALLA_NESTMATES)
			&& (sourceClass->nestHost != destClass->nestHost)
#endif /* defined(J9VM_OPT_VALHALLA_NESTMATES) */
	) {
		result = J9_VISIBILITY_NON_MODULE_ACCESS_ERROR;
	}
...
return result;
```
I therefore assumed that if the class is unsafe or the class is itself, then we have access to private fields. So then if `(J9ROMCLASS_IS_UNSAFE(methodClass->romClass) == 0)` holds, then even with `fieldClass != methodClass` being true, we still have private field access.

If it does not hold hold (ie. `(J9ROMCLASS_IS_UNSAFE(methodClass->romClass) != 0)` - will correct formatting), that is when we require the classes to be the same (ie. `fieldClass != methodClass` must not be true). If the class is not unsafe and the final field is being set by a different class, then the illegal access error setting a final field should be set.

--

I'm still trying to get a test case to run correctly - should I close the PR and re-open it once I have done so?